### ISSUE TITLE
environment_vars always fail config check where they are used (Not in the environment_vars section)

### ISSUE THREAD

**Issue by Petro31** [blakeblackshear/frigate#2484]
### Describe the problem you are having

I don't want support.  This is simply a bug however users aren't allowed to create bugs for some reason.

It appears as if the variables are just not being passed or parsed inside the config.

### Version

Frigate NVR Beta (0.10.0) v 3.2

### Frigate config file

```yaml
environment_vars:
  MY_USER: xxx
  MY_PASSWORD: xxx
  FRIGATE_RTSP_USER: xxx
  FRIGATE_RTSP_PASSWORD: xxx
  RSTP_IP_1: 'xx:xx:xx:xx'
```

later...

```
      - path: rtmp://{RSTP_IP_1}/bcs/channel0_main.bcs?channel=0&stream=0&user={FRIGATE_RTSP_USER}&password={FRIGATE_RTSP_PASSWORD}
```
```


### Relevant log output

```shell
*************************************************************
*************************************************************
***    Your config file is not valid!                     ***
***    Please check the docs at                           ***
***    https://docs.frigate.video/configuration/index     ***
*************************************************************
*************************************************************
***    Config Validation Errors                           ***
*************************************************************
'FRIGATE_RTSP_USER'
*************************************************************
***    End Config Validation Errors                       ***
*************************************************************
```

### FFprobe output from your camera

```shell
does not relate
```

### Frigate stats

```json
does not relate
```

### Operating system

HassOS

### Install method

HassOS Addon

### Coral version

USB

### Network connection

Wired

### Camera make and model

does not relate

### Any other information that may be helpful

No matter what environment variable is used first, that's what will fail the config check.  If I don't use any, the above configuration passes.  The first attempt to use any variable, the configuration fails check.

Using FRIGATE_ or not does not make a difference, it simply does not accept the environment variable.

**Comment by Petro31** [comment]
And if environment vars are not to be used with hassOS, then this section needs to be removed

https://docs.frigate.video/configuration/advanced/#environment_vars

**Comment by blakeblackshear** [comment]
Typically environment vars are set when starting the container with Docker to avoid including them in the config file. The do work that way. If you are setting your RTSP password environment variable in the same config file where you are using it, what are you gaining?

**Comment by Petro31** [comment]
Unless I missed something, I was under the impression that you cannot set environment variables for the HassOS Addon.  Hence why I'm doing it this way.  I would love for another way and I would love for it to be in a separate file.  It doesn't seem like the addon has that ability, which is why I'm going this route.

**Comment by blakeblackshear** [comment]
There isnt a way to set environment variables in the addon configuration page in HassOS? I thought that was there, but I don't have a HassOS instance readily available to check.

I still don't understand why you would bother setting the env variables in the same config file where you are using them. If they are in the same file, I see no security (or other) benefit.

**Comment by Petro31** [comment]
> There isnt a way to set environment variables in the addon configuration page in HassOS? 

There does not appear to be, the configuration page only contains ports for the RTMP streams and Web Interface

> I still don't understand why you would bother setting the env variables in the same config file where you are using them. If they are in the same file, I see no security (or other) benefit.

It's not for security, it's for management.  I have 8 cameras with the same login/password.  I want 1 user, 1 password for easy updated. 

I also share my configuration with many people to help them learn because people want example configurations, they (non technical users) do not want to read documentation.  I currently do not share my frigate yaml because I cannot separate passwords and users from the file and I don't want to take the time to replace my username and password every time I share or every time I have to reorganize the file when the config validation changes (like it did recently).

I would have gladly written all this in myself however you aren't using volupuous and I'd have to learn the validation engine you're using and I don't have the time for that currently (Maybe in the future).

All I do know is that the documents state that you can use environment_vars to create environment variables inside it's configuration.  I cannot create the variables for HassOS addon on my production environment and I'm not going to be swapping my production environment to docker.  So, it seems my hands are tied and I'm forced to use environment_vars, which does not work. 

TLDR:  I want to manage and share my configuration without passwords and use variables as passwords in HassOS Addon on my production machine.

**Comment by blakeblackshear** [comment]
That makes sense and I think I know why they don't work here. There may be a way to allow environment variables to be added to the addon and/or add these in at config load time.

**Comment by Petro31** [comment]
I think it would be a good addition if environment_vars doesn't do what it eludes to doing.  I didn't put much time into investigating how the code works.  I did a half ass search all files for adding the environment_vars to environment variables in python and didn't see it, so I wasn't sure if another method was used.

Either way, I think this is something that many people will use.  Especially people from HA who want secrets in frigate.yaml.  They can use this method instead, killing 2 birds with 1 stone.

**Comment by blakeblackshear** [comment]
It definitely works. This is the first time someone has tried to use it to specify and substitute variables that are defined in the same file. Everyone else is using it to set environment variables for ffmpeg.

**Comment by slepiavka** [comment]
So does this mean that env. variables **can** be passed on to the HASS Add-on for hw acceleration purposes? Haven't found an issue that describes how to yet...

**Comment by blakeblackshear** [comment]
@slepiavka For hw acceleration purposes, you can set them in the frigate config file and the will be set before frigate spawns the ffmpeg sub process.
![image](https://user-images.githubusercontent.com/569905/155240089-cad3f8e4-d6da-4555-ba48-4f34a72e0dfe.png)

**Comment by Strux-DK** [comment]
I'm looking for something like this and can not seem to get the environment variables to work. But then again, they won't do much if i have to set them in the same file. I'm not used to Github, so sorry if this has already been asked and answered, but are you looking at making Frigate support using the secrets.yaml file?

**Comment by NickM-27** [comment]
I personally think the optimal solution would be the ability to just define env vars in the addon configuration screen. Curious if anything has changed around that.

**Comment by wsidl** [comment]
Late to the game here, but as I'm exploring frigate and have a strong dev background with DRY tendencies, I also started using `environment_vars` to set common values that could change and/or used in multiple places.

```yaml
environment_vars:
  CAM_1_HOST: 192.168.x.x
  CAM_1_USER: user
  CAM_1_PWD: password
```

@blakeblackshear, I'll concede if this block is intended for any child processes like ffmpeg, but the [documentation](https://docs.frigate.video/configuration/advanced/#environment_vars) is vague about where these values are used. I have control of my environment, but I also don't want to repeat error-prone values nor do I want to update the container variables without needing to re-deploy the container where a quick service restart from the UI will do. I would even consider YAML anchor/alias's if it not for needing to embed these values into strings.

Now the code in the `dev` branch has been significantly refactored compared to what I'm running today as "latest" but it appears the latest commit by @gtsiam should start allowing these env vars to get adopted and used for substitutions. I'm a bit weary of adding them to the OS's environ (on [line 27](https://github.com/blakeblackshear/frigate/blob/bbbb3b4a06691b335e5b3622545a67f64b41a013/frigate/config/env.py#L27)) and would encourage more flexibility by adding them to their own collection. This means all of your env var sources are distinct and traceable in case you need to use variables from one source over another, or use different orders for overrides.

> Should support Python 3.7+. Switch the dictionary merge syntax on L19 to the more convenient 
> `v.format(**(USER_ENV_VARS | FRIGATE_ENV_VARS))` but would require Python 3.9+

```diff
diff --git a/frigate/config/env.py b/frigate/config/env.py
--- a/frigate/config/env.py
+++ b/frigate/config/env.py
@@ -7,2 +7,3 @@
FRIGATE_ENV_VARS = {k: v for k, v in os.environ.items() if k.startswith("FRIGATE_")}
+ USER_ENV_VARS: dict[str, Any] = {}
# read docker secret files as env vars too
@@ -17,2 +18,2
def validate_env_string(v: str) -> str:
-    return v.format(**FRIGATE_ENV_VARS)
+    return v.format(**{**USER_ENV_VARS, **FRIGATE_ENV_VARS})
@@ -24,6 +25,6
def validate_env_vars(v: dict[str, str], info: ValidationInfo) -> dict[str, str]:
    if isinstance(info.context, dict) and info.context.get("install", False):
        for k, v in v:
-             os.environ[k] = v
+             USER_ENV_VARS[k] = v

    return v
```

**Comment by gtsiam** [comment]
Oh yeah; the work I did in that commit was primarily a refactor to clean up frigate's lifecycle but it should have fixed this, if I'm understanding the issue correctly. @NickM-27 should we close this?

@wsidl I consider the config code to still be very dubious, but it's not a priority for me. Note: I'm not a maintainer. I just took a look at the code like you, went "we can do better" and started sending PRs. So; PRs welcome!

**Comment by NickM-27** [comment]
yes I believe this works as expected now

**Comment by powerbling** [comment]
Sorry for re-opening, I found out that this still doesn't work when substituting camera ips in go2rtc's config.
Is this intended behavior or could this be resolved?
I think it would be useful as in the cases stated above by others, but in the streams section of the config.

e.g.:
```yaml
environment_vars:
  IP_CAMERA_FRONT: 10.0.50.100
  IP_CAMERA_BACK: 10.0.50.199

go2rtc:
  streams:
    front_camera:
      - "rtsp://login:pw@{IP_CAMERA_FRONT}:554/cam/realmonitor?channel=1&subtype=0"
    front_camera_sub:
      - "rtsp://login:pw@{IP_CAMERA_FRONT}:554/cam/realmonitor?channel=1&subtype=1"
    back_camera:
      - "rtsp://login:pw@{IP_CAMERA_BACK}:554/cam/realmonitor?channel=1&subtype=0"
    back_camera_sub:
      - "rtsp://login:pw@{IP_CAMERA_BACK}:554/cam/realmonitor?channel=1&subtype=1"
```
